src/librustc_codegen_llvm/intrinsic.rs

```diff
@@ -731,6 +731,18 @@ impl IntrinsicCallMethods<'tcx> for Builder<'a, 'll, 'tcx> {
                 return;
             }
 
+            "ptr_guaranteed_eq" | "ptr_guaranteed_ne" => {
+                let a = args[0].immediate();
+                let b = args[1].immediate();
+                let a = self.ptrtoint(a, self.type_isize());
+                let b = self.ptrtoint(b, self.type_isize());
```

Comment:
Is there a non-obvious reason you `ptrtoint` before comparing? `icmp` can work fine with pointer types, as long as the pointee types match, which in this case they should.

Reply:
oh. I didn't know that. I basically grabbed this off https://github.com/rust-lang/rust/blob/38bd83df88288f2f8d1fc2dd317189cac3825920/src/librustc_codegen_llvm/intrinsic.rs#L734-L748 which I'm not sure whether it needs that either.

Reply:
> which I'm not sure whether it needs that either.

It does. LLVM supports comparison of pointer types but not subtraction.

Reply:
So confusing. There's probably a good reason for that, but still... So many inconsistencies.

Reply:
FWIW, there's a plan to add a `psub` operation taht directly subtracts pointers. ;)

(There is no pointer addition, either. Instead there is `getelementptr` as the specific instruction to do pointer arithmetic. That makes a lot of sense IMO.)